### PR TITLE
Explode column types and generate converters

### PR DESCRIPTION
**Pull request by pgovind**
First commit of two. Pretty straightforward. This PR explodes the column types and adds converters. The converters are used in the next PR which adds binary operation APIs on the exploded columns.

### CODE REVIEW

File: src/Microsoft.Data.Analysis/Converters.tt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+<#
+                string capitalizedLoopPrimitive = GetCapitalizedPrimitiveTypes(typeLoop.Key);
+#>
+                return (I<#=typeFirstCharUpper#>Converter<T>)new <#=capitalizedLoopPrimitive#><#=typeFirstCharUpper#>Converter();
+            }
+<#
+    }
+#>
+            throw new NotSupportedException();
+
```
Comment (gfoidl) [inline src/Microsoft.Data.Analysis/Converters.tt]: Nit: remove empty line

File: src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.tt [hunk 2]
```diff
@@ -28,10 +28,11 @@ namespace Microsoft.Data.Analysis
         public override PrimitiveDataFrameColumn<bool> <#=method.MethodName#>(DataFrameColumn column)
 <# } #>
         {
-            switch (column)
+<#          // The column here could be a PrimitiveColumn or an exploded column like IntDataFrameColumn#>
+            switch (column.DataType)
             {
 <# foreach (TypeConfiguration type in typeConfiguration) { #>
-                case PrimitiveDataFrameColumn<<#=type.TypeName#>> <#=type.TypeName#>Column:
+                case Type <#=type.TypeName#>Type when <#=type.TypeName#>Type == typeof(<#=type.TypeName#>):
```
Comment (gfoidl) [inline src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.tt]: That's nice 👍

File: src/Microsoft.Data.Analysis/IntDataFrameColumn.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+    public partial class IntDataFrameColumn : PrimitiveDataFrameColumn<int>
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/IntDataFrameColumn.cs]: Int32. We should be using the `System.XXX` type names in APIs. Using the C# names (int, long, short) is confusing to other languages - i.e. F#.

File: src/Microsoft.Data.Analysis/IntDataFrameColumn.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public IntDataFrameColumn(string name, IEnumerable<int?> values) : base(name, values) { }
+
+        public IntDataFrameColumn(string name, IEnumerable<int> values) : base(name, values) { }
+
+        public IntDataFrameColumn(string name, long length = 0) : base(name, length) { }
+
+        public IntDataFrameColumn(string name, ReadOnlyMemory<byte> buffer, ReadOnlyMemory<byte> nullBitMap, int length = 0, int nullCount = 0) : base(name, buffer, nullBitMap, length, nullCount) { }
+
+        internal IntDataFrameColumn(PrimitiveDataFrameColumn<int> intColumn) : base(intColumn.Name, intColumn._columnContainer) { }
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/IntDataFrameColumn.cs]: I don't see this ctor being used. Can we remove it until it is necessary?

Comment (eerhardt) [reply src/Microsoft.Data.Analysis/IntDataFrameColumn.cs]: We would also not need to make `_columnContainer` internal on the PrimitiveDataFrameColumn class.

---
In reply to: [387168118](https://github.com/dotnet/corefxlab/pull/2857#discussion_r387168118) [](ancestors = 387168118)

File: src/Microsoft.Data.Analysis/FloatDataFrameColumn.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+    public partial class FloatDataFrameColumn : PrimitiveDataFrameColumn<float>
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/FloatDataFrameColumn.cs]: `SingleDataFrameColumn`.

File: src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.cs [hunk 6]
```diff
@@ -16,33 +16,33 @@ public partial class PrimitiveDataFrameColumn<T> : DataFrameColumn
     {
         public override DataFrameColumn Add(DataFrameColumn column, bool inPlace = false)
         {
-            switch (column)
+            switch (column.DataType)
             {
-                case PrimitiveDataFrameColumn<bool> boolColumn:
+                case Type boolType when boolType == typeof(bool):
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.cs]: I think your existing pattern matching will work correctly here. Why this change?

Also, for the existing code, you don't need the additional cast below. You can just say:

```C#
            switch (column)
            {
                case PrimitiveDataFrameColumn<bool> boolColumn:
                    return AddImplementation(boolColumn, inPlace);
```

Comment (pgovind) [reply src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.cs]: Right, I thought the same thing. But, we have new column types now, so we need to test against the datatype it holds instead of the column type itself. So `IntDataFrameColumn` and `PrimitiveDataFrameColumn<int>` would both work now

File: src/Microsoft.Data.Analysis/IntDataFrameColumn.cs (hunk 4 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/IntDataFrameColumn.cs]: Yup. Removed in all the new columns

File: src/Microsoft.Data.Analysis/IntDataFrameColumn.cs (hunk 3 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/IntDataFrameColumn.cs]: Done.

File: src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.cs (hunk 6 above)
Comment (eerhardt) [reply src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.BinaryOperations.cs]: I tested this in an isolated environment, and it works:

```C#
    class Primitive<T> where T : struct
    {
    }
    class Int32Column : Primitive<int>
    {
    }
    class Program
    {
        static void Main(string[] args)
        {
            Print(new Int32Column());
            Print(new Primitive<int>());
            Print(new Primitive<bool>());
        }
        private static void Print(object o)
        {
            switch (o)
            {
                case Primitive<int> intColumn:
                    Console.WriteLine("Got int");
                    break;
                case Primitive<long> intColumn:
                    Console.WriteLine("Got long");
                    break;
                default:
                    Console.WriteLine("got nothing");
                    break;
            }
        }
    }
```

prints:
```
Got int
Got int
got nothing
```

File: src/Microsoft.Data.Analysis/UIntDataFrameColumn.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+    public partial class UIntDataFrameColumn : PrimitiveDataFrameColumn<uint>
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/UIntDataFrameColumn.cs]: `UInt32`. Same for ULong and UShort.

File: src/Microsoft.Data.Analysis/ShortDataFrameColumn.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+    public partial class ShortDataFrameColumn : PrimitiveDataFrameColumn<short>
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/ShortDataFrameColumn.cs]: `Int16DataFrameColumn`

File: src/Microsoft.Data.Analysis/Converters.cs [hunk 9]
```diff
@@ -0,0 +1,948 @@
...
+    {
+        public static IDoubleConverter<T> Instance { get; } = DoubleConverter.GetDoubleConverter<T>();
+    }
+    internal static class DoubleConverter
+    {
+        public static IDoubleConverter<T> GetDoubleConverter<T>()
+        {
+            if (typeof(T) == typeof(byte))
+            {
+                return (IDoubleConverter<T>)new ByteDoubleConverter();
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/Converters.cs]: (nit) is the cast necessary here?

File: src/Microsoft.Data.Analysis/ColumnArithmeticTemplate.ttinclude [hunk 10]
```diff
@@ -38,6 +38,56 @@
         return $"{keyword} (typeof(T) == typeof({type.TypeName}))";
     }
 
+        // A way to discern implicit conversions. For ex: short has primitivitty 2. int has primitivitty 3. primitivitty(short) + primitivitty(int) > 2 * primitivitty(short) implying that a conversion has to take place
```
Comment (eerhardt) [inline src/Microsoft.Data.Analysis/ColumnArithmeticTemplate.ttinclude]: Are these changes necessary/intended?

File: src/Microsoft.Data.Analysis/Converters.cs (hunk 9 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/Converters.cs]: Yup, an implicit conversion does not exist between IDoubleConverter<T> and ByteDoubleConverter (even though it implements IDoubleConverter<byte>)

File: src/Microsoft.Data.Analysis/ColumnArithmeticTemplate.ttinclude (hunk 10 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/ColumnArithmeticTemplate.ttinclude]: Yup. `Converters.tt` uses this map to decide which APIs to generate.
